src/rust/engine/src/externs/interface.rs

```diff
@@ -1671,16 +1672,26 @@ fn write_digest(
     }
 
     block_in_place_and_wait(py, || async move {
-      core
-        .store()
+      let store = core.store();
+      store
         .materialize_directory(
           destination.clone(),
-          lifted_digest,
+          lifted_digest.clone(),
           true, // Force everything we write to be mutable
           &BTreeSet::new(),
           fs::Permissions::Writable,
         )
-        .await
+        .await?;
+
+      if invalidate {
+        let snapshot = store::Snapshot::from_digest(store, lifted_digest).await?;
+        let files = snapshot.tree.files(SymlinkBehavior::Aware);
+        py_scheduler
+          .0
+          .invalidate_paths(&files.into_iter().collect());
```

Comment:
I notice that this logs at `info` level so every `pants fmt ::` call gets a message like: `10:18:52.27 [INFO] external invalidation: cleared 85 and dirtied 2 nodes for: {"file57.py", "file79.py", "file32.py", ..., "file100.py", "file56.py"}`, which seems unhelpful. Any objections to downgrading that to `debug!`, or alternatives?

https://github.com/pantsbuild/pants/blob/de1a3d32862f65fe5dde73afa1aa9d341d05077b/src/rust/engine/src/context.rs#L696-L699

Reply:
This message is very useful when it is triggered in the background in `pantsd`. So I think that some caller-dependent subtlety will need to be introduced.

Reply:
Is there a way to log at info level to the pants.log file but not to the UI?

Or maybe truncate it in the UI but not log file?

Reply:
> Is there a way to log at info level to the pants.log file but not to the UI?

Yes, but only on a thread-by-thread basis currently: individual calls can't be directed to various places. See `stdio_thread_set_destination` (...currently unused).

Reply:
Hm, sorry, I'm still not sure of the right way to handle this. Are you thinking I could bracket the invalidation call with the appropriate set of `stdio::set_thread_destination`/`stdio::get_destination` calls, with a `Destination` pointing to `pants.log`?

That feels... very fiddly to me.

Making the level conditional be passed in by the caller seems a bit more direct (so the foreground call would log at debug and the daemon calls would log at info), although will require some plumbing.

Reply:
I've done that plumbing, switching the existing `caller` arg over to be an enum since there's only two values (notify and external), and keying the log level off that. AFAIK, the only other caller of the external invalidation is tests, so I've just downgraded all of them to debug.

https://github.com/pantsbuild/pants/blob/dda7df1cd0ef019fce54e453ea6ea1124fbadeb0/src/python/pants/testutil/rule_runner.py#L487

Reply:
That looks great: thanks!